### PR TITLE
0x trading testing script

### PR DESCRIPTION
**Pull request by haythem96**
# Task: 0x trading testing script

## High Level Description

This PR add a script to test Gamma actions batched with 0x trading on Kovan testnet.
The script use two address:
- Maker: account[1] from the imported wallet that will sign a 0x order to sell 1 USDC for 1 Call option
- Taker: account[0] from the imported wallet that will mint a Call option and sell it using the Trade0x Callee

The Taker account need to have at least 1USDC and few ETH to approve USDC transfer.
The Maker account need to have at least 1ETH to collateralize the minted option + few ETH for 0x and transaction fees.

Also this PR add a bunch of scripts that help interacting with the protocol:
- Deploy new Permit Callee
- Deploy new 0x Callee
- Set call restriction in Controller
- Whitelist callee address

**Usage:**
`truffle exec scripts/trade0x-kovan.js --network kovan --makerPrivateKey 0x0000...`
where `makerPrivateKey` is the account private key used to sign 0x order.

**Output:**
```
🔥🔥🔥 STARTING 0x TRADING SCRIPT 🔥🔥🔥
0x order signer address: 🖋️  0x543361Be968e2f59087316514eaee9e9E9A30aF7
Taker address:  0x8c8b4C5caEAf1A6FB8308a6336932B3e3419704C
WETHUSDC 30-January-2021 640Call WETH Collateral 0x10980E19f0b40Abde86Df23aFE7Fb90DF74C7D0d
Signed 0x order with signature: 🖋️  0x1c46f91a51cc20e17a91b00fb32a3ae7b0d690b74b0d287a11a787f327db33761e0a486220fbf0eb91d3138f4c37ae680d05c18b2d5278f566b1e825651af2814d02
ETH needed: 💰  1003500000000000000
MarginPool WETH balance before: 📊  25408108658113122978
Maker USDC balance before: 📊  98000000
Maker Otoken balance before: 📊  200000000
Taker USDC balance before: 📊  941437399
Taker Otoken balance before: 📊  500000000
Otoken total supply before: 📊  700000000
Approving 0x Callee and calling Operate 🚀
MarginPool WETH balance after: 📈  26408108658113122978
Maker USDC balance after: 📉  97000000
Maker Otoken balance after: 📈  500000000
Taker USDC balance after: 📈  942437399
Taker Otoken balance after: 📉 500000000
Otoken total supply after: 📈  800000000
```

### Code

- [x] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### Documentation

- [ ] Is your code up to date with the spec? 
- [ ] Have you added your tests to the testing doc?

### CODE REVIEW

**Comment by antoncoding** [comment]
@haythem96 can you rebase your branch to the current master? It's been containing commit message from last Dec 😂

File: scripts/trade0x-kovan.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  console.log("MarginPool WETH balance before: 📊", marginPoolWethBalanceBefore.toString())
+  console.log("Maker USDC balance before: 📊", makerUsdcBalanceBefore.toString())
+  console.log("Maker Otoken balance before: 📊", makerCall1BalanceBefore.toString())
+  console.log("Taker USDC balance before: 📊", user1UsdcBalanceBefore.toString())
+  console.log("Taker Otoken balance before: 📊", user1Call1BalanceBefore.toString())
+  console.log("Otoken total supply before: 📊", oTokenSupplyBefore.toString())
+
+  console.log("Approving 0x Callee and calling Operate 🚀")
+
+  await callOption1.approve(trade0xCallee.address, optionsToMint, {from: taker})
```
Comment (aparnakr) [inline scripts/trade0x-kovan.js:242]: I thought we were going to use the permit callee for this approval?

File: scripts/trade0x-kovan.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+  const latestBlock = new BigNumber((await web3.eth.getBlock('latest')).timestamp);
+  let expiry = new BigNumber(latestBlock).plus(expiryTime)
+  let expiryDate = new Date(expiry.toNumber() * 1000)
+  expiryDate.setUTCHours(8)
+  expiryDate.setUTCMinutes(0)
+  expiryDate.setUTCSeconds(0)
+  expiry = new BigNumber(Math.floor(expiryDate.getTime() / 1000))
+
+  let callOption1;
```
Comment (aparnakr) [inline scripts/trade0x-kovan.js:107]: can we also test put options? :D

**Review by aparnakr — APPROVED** [review]

File: scripts/trade0x-kovan.js (hunk 1 above)
Comment (haythem96) [reply scripts/trade0x-kovan.js:242]: The Otoken in Kovan currently have no `permit()`, so I just used the `approve()`

File: scripts/trade0x-kovan.js (hunk 2 above)
Comment (haythem96) [reply scripts/trade0x-kovan.js:107]: IMO it is the same as the Trade0x is agnostic to what the option type is, it is the same asset/contract.

File: scripts/trade0x-kovan.js [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+  // create 0x order
+  const order = createOrder(
+    EXCHANGE_ADDR,
+    makerEtherJS.address,
+    usdc.address,
+    callOption1.address,
+    new BigNumber(createTokenAmount(1, 6)),
+    new BigNumber(optionsToMint),
+    42
```
Comment (antoncoding) [inline scripts/trade0x-kovan.js:136]: better make this 42 a constant along with all the address `consts`, so might be easier to reuse this script in the future.

File: test/unit-tests/trade0xCallee.test.ts [hunk 4]
```diff
@@ -63,6 +63,7 @@ contract('Trade0xCallee', ([payableProxy, taker, staking, random]) => {
       token2.address,
       new BigNumber(token1Amount),
       new BigNumber(createTokenAmount(100, 8)),
+      1,
```
Comment (antoncoding) [inline test/unit-tests/trade0xCallee.test.ts:66]: same here, use `networkId` instead of `1` would be helpful.

**Review by antoncoding — APPROVED** [review]

**Pull request merged**